### ISSUE TITLE
Zeroshot predictions giving same output regardless of the input signal

### ISSUE THREAD

**Issue by akhilpm** [Mobile-Sensing-and-UbiComp-Laboratory/NormWear#8]
I was testing you model on ECG data (1 channel). Using 5 or 6 seconds window, I tried to predict the cognitive load of the subject. I tried several prompts, but here is a sample:

task_of_interest = [
    "is this subject in high cognitive load state?"
]

options = [
    "The subject is in high cognitive load",
    "The subject is having low cognitive load",
]

Regardless of the prompts or window size, I am getting predictions like below always (the model is predicting one class regardless of the input)

281:   GT: 0 Pred: [1]   prob: tensor([[0.4964, 0.5036]], device='cuda:0')
282:   GT: 0 Pred: [1]   prob: tensor([[0.4960, 0.5040]], device='cuda:0')
283:   GT: 0 Pred: [1]   prob: tensor([[0.4965, 0.5035]], device='cuda:0')
284:   GT: 0 Pred: [1]   prob: tensor([[0.4959, 0.5041]], device='cuda:0')
285:   GT: 0 Pred: [1]   prob: tensor([[0.4960, 0.5040]], device='cuda:0')
286:   GT: 1 Pred: [1]   prob: tensor([[0.4963, 0.5037]], device='cuda:0')
287:   GT: 1 Pred: [1]   prob: tensor([[0.4948, 0.5052]], device='cuda:0')
288:   GT: 1 Pred: [1]   prob: tensor([[0.4968, 0.5032]], device='cuda:0')
289:   GT: 1 Pred: [1]   prob: tensor([[0.4966, 0.5034]], device='cuda:0')
290:   GT: 1 Pred: [1]   prob: tensor([[0.4955, 0.5045]], device='cuda:0')
291:   GT: 1 Pred: [1]   prob: tensor([[0.4964, 0.5036]], device='cuda:0')
292:   GT: 1 Pred: [1]   prob: tensor([[0.4957, 0.5043]], device='cuda:0')
293:   GT: 1 Pred: [1]   prob: tensor([[0.4959, 0.5041]], device='cuda:0')
294:   GT: 1 Pred: [1]   prob: tensor([[0.4958, 0.5042]], device='cuda:0')

Can the authors please help me to understand the issue?

**Comment by akhilpm** [comment]
I also tried the pre-processing technique you mentioned in the paper. It is implemented as

def preprocess_signal(x, original_sr=120, target_sr=65, window_sec=6, sigma_sec=0.02):
    """
    x: Tensor of shape [1, 1, 7200]  (batch, channel, time)
    Returns a tensor of shape [1, 1, target_sr * window_sec]
    """
    # Step 1: Convert torch tensor to numpy
    x_np = x.squeeze().cpu().numpy()  # shape: [7200]
    # Step 2: Detrend
    x_np = detrend(x_np)
    # Step 3: Gaussian smoothing
    sigma_samples = int(sigma_sec * original_sr)  # ~0.02s * 120Hz = ~2-3 samples
    kernel_size = sigma_samples * 6  # cover +-3σ range
    if kernel_size % 2 == 0:
        kernel_size += 1  # make it odd
    gauss_kernel = gaussian(kernel_size, std=sigma_samples)
    gauss_kernel /= gauss_kernel.sum()  # normalize
    x_np = convolve(x_np, gauss_kernel, mode='same')
    # Step 4: Resample to 65 Hz
    target_length = int(target_sr * (len(x_np) / original_sr))
    x_np = resample(x_np, target_length)
    # Step 5: Take 6-second chunks (390 samples at 65 Hz)
    if len(x_np) >= target_sr * window_sec:
        x_np = x_np[:target_sr * window_sec]  # keep first 6 seconds
    else:
        # Pad if shorter
        pad_len = target_sr * window_sec - len(x_np)
        x_np = np.pad(x_np, (0, pad_len), mode='constant')
    # Step 6: Back to tensor
    x_out = torch.tensor(x_np, dtype=torch.float32).unsqueeze(0).unsqueeze(0)  # shape: [1,1,N]
    return x_out

**Comment by yunfeiluo** [comment]
Hi, 

After reviewing the released code and checkpoints, we believe the most likely explanation is a mismatch between the backbone checkpoint and the MSiTF checkpoint. The MSiTF pretraining pipeline is available at: https://github.com/Mobile-Sensing-and-UbiComp-Laboratory/NormWear/tree/main/zero_shot which has a similar workflow as the pretrain pipeline for backbone model. During the research phase of this project, there are multiple model variants that may come from different stages of development, including hyperparameter tuning and ablation experiments conducted during the research and revision process. The final released checkpoints are more for a reference. Thus, we recommend to re-train the MSiTF with the backbone checkpoint being controlled for research purposes. 

We would suggest first checking whether the model is in evaluation mode and whether inference is performed under torch.no_grad(). Another suggestion is to retrain the MSiTF module locally on top of a backbone checkpoint, either from the GitHub repository or from the Hugging Face release. You may also consider trying pretraining with our newer model architecture with the latest [NormWear 2](https://huggingface.co/mosaic-laboratory/normwear2) checkpoints. This would ensure that the backbone and MSiTF checkpoints are matched correctly.

Again, these are only recommendations, as the context may differ in several ways, such as varied data distribution and differences in the specific focus of the research study. The zero-shot experiments in the NormWear-1 paper were intended as a preliminary exploration of the feasibility of zero-shot inference rather than to represent the full potential of the framework in downstream applications. From a methodological perspective, we also believe there is considerable room for improvement, and that more effective modeling strategies than the one we proposed remain to be explored as briefly discussed in the paper.  

Best regards,
NormWear Team
